File: backend/src/main/scala/slaydemo/backend/battle/objects/BattleResultRecord.scala
package slaydemo.backend.battle.objects

import slaydemo.backend.identity.objects.{DisplayName, PlayerHandle}

final case class BattleResultRecord(
  battleId: BattleId,
  handle: PlayerHandle,
  displayName: DisplayName,
  finishedAt: EpochMillis,
  finishedAtLabel: String,
  durationMs: DurationMillis,
  score: Score,
  placement: Option[BattlePlacement],
  survivalOutcome: BattleSurvivalOutcome,
  ratingBefore: Rating,
  ratingDelta: RatingDelta,
  ratingAfter: Rating,
  resultLabel: BattleResultLabel,
  modeLabel: BattleModeLabel,
  mapLabel: BattleMapLabel,
  highlightLine: BattleHighlightLine,
  playersLine: BattlePlayersLine,
  timelineHint: BattleTimelineHint,
  currentLoadout: Option[String]
) {
  def resultId: BattleResultId =
    BattleResultRecord.resultId(battleId, handle)

  def aliveAtEnd: Boolean =
    BattleSurvivalOutcome.aliveAtEnd(survivalOutcome)
}

object BattleResultRecord {
  def resultId(battleId: BattleId, handle: PlayerHandle): BattleResultId =
    BattleResultId(s"${battleId.value.trim}:${handle.key}")
}

File: frontend/src/objects/battle/microservices/results/objects/result/BattleResultRecord.ts
import type {
  BattleId,
  BattleMapLabel,
  BattleModeLabel,
  DurationMillis,
  EpochMillis
} from "../../../../objects/core/BattleCoreScalars";
import type { BattleSurvivalOutcome } from "../../../actors/objects/player/BattleSurvivalOutcome";
import type { Rating } from "../../../actors/objects/player/BattlePlayerRating";
import type { Score } from "../../../actors/objects/player/BattlePlayerStats";
import type {
  BattleHighlightLine,
  BattlePlacement,
  BattlePlayersLine,
  BattleResultLabel,
  BattleTimelineHint,
  RatingDelta
} from "./BattleResultPresentationValues";

export interface BattleResultRecord {
  battleId: BattleId;
  handle: string;
  displayName: string;
  finishedAt: EpochMillis;
  finishedAtLabel: string;
  durationMs: DurationMillis;
  score: Score;
  placement: BattlePlacement | null;
  survivalOutcome: BattleSurvivalOutcome;
  ratingBefore: Rating;
  ratingDelta: RatingDelta;
  ratingAfter: Rating;
  resultLabel: BattleResultLabel;
  modeLabel: BattleModeLabel;
  mapLabel: BattleMapLabel;
  highlightLine: BattleHighlightLine;
  playersLine: BattlePlayersLine;
  timelineHint: BattleTimelineHint;
  currentLoadout: string | null;
}

export function battleResultRecordId(value: Pick<BattleResultRecord, "battleId" | "handle">): string {
  return `${value.battleId.trim()}:${value.handle.trim().toLowerCase()}`;
}

